source/guides/architecture/interactors.md

```diff
@@ -0,0 +1,382 @@
+because user sign-up is not trivial.
+
+
+# Background
+Passwords should **never** be stored in plain-text.
+
+We should first _hash_ the password.
+This is a function that always returns the same random output for the
+same input.
+It is _one-way_,
+which means it's extremely hard to reverse it,
+that is, to get the password from the hashed output.
+(This output is called a digest.)
+
+We don't store plain-text passwords because if our database is compromised,
+we don't want to reveal our users' passwords
+(since many people re-use passwords).
+
+Instead, attackers would only get the hashed results.
+
+Note: this approach is still vulnerable to
+[Rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table).
+
+In order to combat this, we'll use a hashing scheme that employs a
+[`salt`](https://en.wikipedia.org/wiki/Salt_(cryptography)).
+
+# Preparing
+Let's say we have our `bookshelf` application,
+from the [Getting Started]((/guides/getting-started)
+and we want to add a feature.
+
+Clone the `bookshelf` application so we're starting from the same place.
+```bash
+git clone git@github.com:hanami/bookshelf.git hanami-bookshelf
+cd hanami-bookshelf
+bundle install
+HANAMI_ENV=development hanami db prepare
+HANAMI_ENV=test hanami db prepare
+bundle exec rake # All the tests should pass
+```
+
+# Adding User model
+```bash
+hanami generate model User
+```
+
+You'll see an entity (`User`) and a repository (`UserRepository`) created,
+as well as their tests files.
+
+You'll also see a migration was generated,
+something like:
+`db/migrations/20170326152126_create_users.rb`
+
+Edit this file:
+
+```ruby
+Hanami::Model.migration do
+  change do
+    create_table :users do
+      primary_key :id
+      string :email, null: false, unique: true
+      string :password_digest
+
+      column :created_at, DateTime, null: false
+      column :updated_at, DateTime, null: false
+    end
+  end
+end
+```
+
+We'll add three columns:
+- `email`, we require this field, and make sure it's unique
+- `password_digest`, this will be the user's hashed password
+
+
+Let's create that table:
+
+```bash
+bundle exec hanami db prepare
+```
+
+# Interactor
+Let's create a place for our Interactor to go:
+
+```bash
+mkdir lib/bookshelf/interactors
+mkdir spec/bookshelf/interactors
+```
+
+We put them in `lib/bookshelf` because they're decoupled from the web:
+later you may want to allow users to sign up via an API or an admin portal.
+
+
+Let's add some boilerplate to a new `lib/bookshelf/interactors/sign_up_user.rb` file:
+
+```ruby
+require 'hanami/interactor'
+
+class SignUpUser
+  include Hanami::Interactor
+
+  def initialize
+
+  end
+
+  def call
+
+  end
+end
+```
+
+These are the only two public methods this class has:
+`initialize`, to set-up the data, and
+`call` to actually perform the operation(s).
+
+
+Likewise, we need to fill out a new spec: `spec/bookshelf/interactors/sign_up_user_spec.rb`
```

Comment:
bookshelf guide is test-first. so the test could be created before the actual class.